KYRA: (EOB) - HP per level rule enahncements

vrza opened the pull request:
More faithful AD&D rules for Eye of the Beholder 1 and 2, according to the discussion in PR #5946.

vrza commented:
It's not a bad idea, I was actually thinking along a similar path...

The total HP for a character is a sum in the form:

HP = h1 / n + a1/n + h2/n + a2/n + ...

hn - Hit Die n (random roll on level up)
an - adjustment n (static value)
n - number of sub-classes (static value 1, 2 or 3)

Since all the components in the sum are integers divided by n, we can calculate HP as

HP = (h1 + a1 + h2 + a2 + ...) / n

So perhaps we can simply store this sum of all Hit Die rolls and Adjustments (a single uint16 is more than enough), and on level up we add to it and divide by n to get current HP.

athrxx commented:
Yes, that should work well. Let's go for that.

I think we should always update the new var, regardless of the checkbox setting. Just not apply it if the option is off. So that you still get sane values if you enable or disable the option somewhere along the way.

Do you want to update the PR with the changes?

For the savegame code, you'll need to increase the version at the top of `saveload.cpp` and add a version check in `EoBCoreEngine::loadGameState()` for reading the new var, so that it will still load older saves.

vrza commented:
> I think we should always update the new var, regardless of the checkbox setting. Just not apply it if the option is off. So that you still get sane values if you enable or disable the option somewhere along the way.

Yeah, this means we should always run the "enhanced" code path to get this sum value (still struggling to think of a good name for it, naming is hard!), regardless of the checkbox.

 If a loaded older-version save game is missing this sum value, on next level up we could fill it in by using hitPointsMax * n.
 
This will need a change up the stack in `EoBCoreEngine::increaseCharacterLevel`, yeah I think I could make this change.

vrza commented:
On an unrelated note, ScummVM save game format is now quite different from original EotB 2 save game format. Perhaps there should be a tool to convert the ScummVM save to original format, so that it can be loaded in the original game, and imported into Eotb 3 (which uses a different engine, not yet supported by ScummVM).

athrxx commented:
> On an unrelated note, ScummVM save game format is now quite different from original EotB 2 save game format. Perhaps there should be a tool to convert the ScummVM save to original format, so that it can be loaded in the original game, and imported into Eotb 3 (which uses a different engine, not yet supported by ScummVM).

You can do it from the debugger console (IIRC Alt+shift+ d or Ctrl+shift+d). I am not at my PC right now, so I am not sure about the exact incantation... Then just type 'help', it will show you the command for creating a DOS save game.

vrza commented:
Very cool, and works, thanks for the pointer!

vrza commented:
I made the changes, the new field is called `hitPointsDividend`. 

Figuring out how to best test, since the logic executes only on level up events.

athrxx commented on engines/kyra/gui/saveload_eob.cpp:
```diff
@@ -85,6 +85,8 @@ Common::Error EoBCoreEngine::loadGameState(int slot) {
 		for (int ii = 0; ii < 3; ii++)
 			c->experience[ii] = in.readUint32BE();
+		if (header.version >= 22)
+			c->hitPointsDividend = in.readSint16BE();
```
If you write it after hitPointsMax, you also have to read it at that position?

vrza replied:
Good catch, corrected.

athrxx commented:
> I made the changes, the new field is called `hitPointsDividend`.
> 
> Figuring out how to best test, since the logic executes only on level up events.

Thanks! I guess, the best method is throwing lots of xp at your party. Like add `points *= 100;` (or even more) in `EoBCoreEngine::increasePartyExperience()`. I think I did it quite like that when I tested certain platform versions, like EOB1 PC-98 or Amiga...

vrza commented:
> I think we should always update the new var, regardless of the checkbox setting

I just realized the challenge here is to roll the Hit Die once, but run both original and enhanced logic with that roll.

athrxx commented:
> > I think we should always update the new var, regardless of the checkbox setting
> 
> I just realized the challenge here is to roll the Hit Die once, but run both original and enhanced logic with that roll.

Yes, true. Pity, since you made such a nice layout :-P.
But maybe it has to go into one function again...

vrza commented:
Maybe, or we can extract function to roll Hit Die for a character class (Hit Die roll is a llegit AD&D concept) and pass the result to both functions.

Speaking of refactoring, I'm looking at this:

```cpp
int hpInc = generateCharacterHitpointsByLevel(charIndex, 1 << levelIndex);
```

but `generateCharacterHitpointsByLevel` will in turn loop and skip everything else:

```	for (int i = 0; i < 3; i++) {
		if (!(levelIndex & (1 << i)))
			continue;
``

Seems like the loop is unnecessary and we could've just direclty used `i = levelIndex`, or am I missing something?

athrxx commented:
I think you're right. It is never really used as a flag.
In `CharacterGenerator::generateStats()` it is also used in the same way.
That function should probably also get the enhancement, so that the stats from the character generation also profit from the new checkbox option.

vrza commented:
Pushed changes:

- got rid of bogus levelIndex flag
- extracted rollHitDie method
- updated character generator

Need review and testing in order to polish.

athrxx commented:
Thanks, looking good. I'll do some testing...

athrxx commented:
> Made hitPointsDividend update to match HP when user changes the HP stat manually in the character generation screen.
> 
> Btw, for Eye of the Beholder II, the checkbox doesn't show up (at least in my build), can you please check that as well?

That's kind of a flaw in the way our launcher options work. If you detect the game again you should get the option right away. But if an option is added at a later point for an already configured target, you have to run the game once, so that the options get upgraded. So, just run EOBII and return to the launcher, the checkbox should now be there.

vrza commented:
Oh, I didn't know that, thanks for sharing that gotcha. Made the code prettier (polished comments, removed unneeded variables,  gave variables like `m` more descriptive names ....).

BJNFNE commented:
Please for all your other commits also the KYRA: prefix like you did for the first one example: KYRA: commit message.

athrxx commented:
Could you please fix the commit messages? They need the `KYRA:` prefix. The first commit is fine, but you forgot it for the other ones.

vrza commented:
Certainly. I squashed them into a single commit with a detailed message, and rebased to latest master.

On an unrelated topic, what do you think about adding additional key bindings -- concretely, QWEASD -- as alternative keys for moving around? These keys don't seem to overlap with any existing key bindings.

athrxx commented:
> Certainly. I squashed them into a single commit with a detailed message, and rebased to latest master.
> 

Thanks! 


> On an unrelated topic, what do you think about adding additional key bindings -- concretely, QWEASD -- as alternative keys for moving around? These keys don't seem to overlap with any existing key bindings


This topic used to come up a lot, before we added the key mapper. Now, users are supposed to use that to have the key bindings they want.

athrxx commented:
Okay, I have hacked up a crude test which generates a buffer of 64 characters (with a preset rand seed), levels them up to max and saves that buffer for comparison. Unfortunately, the memcmp between before and after your patch fails. Which means I have to dig a bit deeper...

vrza commented:
I guess you're using exact same dice roll results when leveling both sets of characters?

vrza commented:
Did you account for character structs having the new hpDividend feed after the patch?

athrxx commented:
> Did you account for character structs having the new `hitPointDividend` field after the patch?

No, I have to fix that. Would be good if that was all to it...

athrxx commented:
Okay, that didn't solve it.
But I see something else. We now always do the rollHitDie() call. Which advances the rand seed.
With the old code, that didn't happen on all code paths.
So, that will mess up the test even if the new code is totally valid...

vrza commented:
That's true. You can try leveling up to some level where dice are alwyas rolled for all character classes (e.g. to level 8).

vrza commented:
Part of the reason why I extracted method `rollHitDie`  was to cleanly separate this non-determinism from the "business logic", makes it easier to test logic.

vrza commented:
Just noticed a bug where I inadvertently moved the character level increase to after calculating Hit Points, fixing it now.

I'll also refactor to extract the check if we need to roll Hit Die.

athrxx commented:
I abs

> Just noticed a bug where I inadvertently moved the character level increase to after calculating Hit Points, fixing it now.
> 
> I'll also refactor to extract the check if we need to roll Hit Die.

No need to worry about that rollHitDie placement. I can move that around for my test...

athrxx commented:
Okay, now, after carefully moving the rollHitDie() a bit, the test is successful if I only do the character generation, without the level up. Now, I have to try that again..

vrza commented:
Pushed change, now new code should be making the same number of RNG calls as old code.

athrxx commented:
> Pushed change, now new code should be making the same number of RNG calls as old code.

Is the that change also needed for the chargen code? Or does it always call it in there, anyway?

vrza commented:
EotB 1 generates new character up to level 3, EotB up to level 7, so in those games Hit Die will be rolled for each level achievable in chargen.

athrxx commented:
Okay, finally, with your latest update (and moving the hitDieRoll call out of the way in chargen) my test succeeds with leveling up to max.
2685 rollDice calls...

So, your code seems perfectly fine.

vrza commented:
Even after 20+ years of programming, that's still nice to hear :)

athrxx commented:
> Even after 20+ years of programming, that's still nice to hear :)

Yeah, in cases like this, it is more about messing up some vars or some typo (and it isn't so easy to see if you mixed an `i` with an `l` or whatever). And I don't think anyone will ever look at that code again (like nobody did for the last 13 years), so it should be as flawless as we can get it.

vrza commented:
Agree. It's always easy to miss something, even if you read each and every line..Having good refactoring technique helps. Testing helps (would be cool to have unit tests for game logic).

I just noticed another interesting thing in the original game logic. When getting the constitution Hit Point adjustment, `constitutionCur` is taken into account and not `constitutionMax`. Not sure if there any items and/or effects that can change the constitution stat, but it's peculiar.

vrza commented:
But I see nothing in the AD&D rules that contradicts that so I guess it's okay.

Ah yes, and what do you think about adding the checkbox to the in-game Preferences menu?

athrxx commented:
Is it ready to merge or did you still want to add something?

> Agree. It's always easy to miss something, even if you read each and every line..Having good refactoring technique helps. Testing helps (would be cool to have unit tests for game logic).
> 
> I just noticed another interesting thing in the original game logic. When getting the constitution Hit Point adjustment, `constitutionCur` is taken into account and not `constitutionMax`. Not sure if there any items and/or effects that can change the constitution stat, but it's peculiar.

Interesting. I quickly checked disasm if that is a bug of mine, but it isn't... Anyway, I am pretty sure that nothing can change the stat in the EOB games. The scripts are very limited in what they can do.


Saving and loading works as it should and I also tried the party transfer (which is also basically some modified and more elaborate way of loading a savegame). All seems good...

Should we merge the PR or did you still want to add anything?

athrxx commented:
> But I see nothing in the AD&D rules that contradicts that so I guess it's okay.
> 
> Ah yes, and what do you think about adding the checkbox to the in-game Preferences menu?

Probably not worth it, with the hacking up of the menus and making up custom strings for all language versions, including Japanese and Chinese (which IIRC doesn't even use a known encoding).
If it made much sense to change it during gameplay that would be a thing to consider. But it isn't even a good idea to mess up the character development by having the option on for one level-up and off for the next. So I guess I like it better to just have it in the launcher (where it will also get properly translated).

vrza commented:
> Should we merge the PR or did you still want to add anything?

We can merge it. All the functionality we discussed is in, and you tested it thoroughly, so it's a good working checkpoint.

athrxx commented:
Thanks!
